src/rgw/driver/d4n/rgw_sal_d4n.h

```diff
@@ -20,29 +20,46 @@
 #include "rgw_oidc_provider.h"
 #include "rgw_role.h"
 #include "common/dout.h" 
+#include "rgw_aio_throttle.h"
 
+#include "rgw_redis_driver.h"
 #include "driver/d4n/d4n_directory.h"
-#include "driver/d4n/d4n_datacache.h"
+#include "driver/d4n/d4n_policy.h"
+
+#include <boost/asio/io_context.hpp>
+#include <boost/redis/connection.hpp>
 
 namespace rgw { namespace sal {
 
 class D4NFilterDriver : public FilterDriver {
   private:
-    RGWBlockDirectory* blk_dir;
-    cache_block* c_blk;
-    RGWD4NCache* d4n_cache;
+    boost::redis::connection* conn;
+    rgw::cache::CacheDriver* cacheDriver;
+    rgw::d4n::ObjectDirectory* objDir;
+    rgw::d4n::BlockDirectory* blockDir;
+    rgw::d4n::CacheBlock* cacheBlock;
+    rgw::d4n::PolicyDriver* policyDriver;
 
   public:
-    D4NFilterDriver(Driver* _next) : FilterDriver(_next) 
+    D4NFilterDriver(Driver* _next, boost::asio::io_context& io_context) : FilterDriver(_next)
     {
-      blk_dir = new RGWBlockDirectory(); /* Initialize directory address with cct */
-      c_blk = new cache_block();
-      d4n_cache = new RGWD4NCache();
+      conn = new boost::redis::connection{io_context};
+
+      rgw::cache::Partition partition_info;
+      partition_info.location = "RedisCache"; // figure out how to fill rest of partition information -Sam
+
+      cacheDriver = new rgw::cache::RedisDriver(*static_cast<boost::redis::connection*>(conn), partition_info); // change later -Sam
+      objDir = new rgw::d4n::ObjectDirectory();
+      blockDir = new rgw::d4n::BlockDirectory();
+      cacheBlock = new rgw::d4n::CacheBlock();
+      policyDriver = new rgw::d4n::PolicyDriver("lfuda");
     }
     virtual ~D4NFilterDriver() {
-      delete blk_dir; 
-      delete c_blk;
-      delete d4n_cache;
+      delete cacheDriver;
+      delete objDir; 
+      delete blockDir; 
+      delete cacheBlock;
+      delete policyDriver;
```

Comment:
it shouldn't be necessary to heap-allocate all this stuff. just make them non-pointer member variables and call their constructors in the member initializer list:
```c++
D4NFilterDriver(Driver* _next, boost::asio::io_context& io_context)
  : FilterDriver(_next),
    conn(boost::asio::make_strand(io_context)),
    ...
```